qemu/cpu/cpu.py

```diff
@@ -1219,20 +1221,42 @@ def patch_arch_init_header(src, target_name):
         lines = f.readlines()
 
     index = 0
-    str_qemu_arch = "    %s = " % qemu_arch
-    found_arch_all = False
+    first_enum_found = False
+    arch_definitions = {}
 
     for i, line in enumerate(lines):
-        if found_arch_all:
-            if str_qemu_arch in line:
-                return
-            elif line == "};\n":
-                lines.insert(i, "%s(1 << %d),\n" % (str_qemu_arch, index))
+        if first_enum_found:
+            if line == "};\n":
+                index = i
                 break
-            else:
-                index += 1
-        elif line == "    QEMU_ARCH_ALL = -1,\n":
-            found_arch_all = True
+            arch_e_d_match = re_arch_enum_definition.search(line)
+            if arch_e_d_match:
+                arch_definitions[arch_e_d_match.group(1)] = (
+                    int(arch_e_d_match.group(2))
+                )
+        if line == "enum {\n":
+            first_enum_found = True
+
+    if qemu_arch in arch_definitions:
+        return
+
+    QEMU_ARCH_NONE_val = None
+    if "QEMU_ARCH_NONE" in arch_definitions:
+        # subtracts 2 because QEMU_ARCH_NONE is separated by an empty line
+        index = index - 2
+        QEMU_ARCH_NONE_val = arch_definitions.pop("QEMU_ARCH_NONE")
+
+    arch_val = max(arch_definitions.values()) + 1
+
+    if arch_val == QEMU_ARCH_NONE_val:
```

Comment:
Is QEMU_ARCH_NONE absent in some Qemu versions and is 31 bit free to use for such a version?

Else, u can just `if arch_val > 30:` below and remove this condition.

Reply:
Anyway, a comment should clarify this.

Reply:
```
if arch_val == QEMU_ARCH_NONE_val or if arch_val > 31:
    raise ...
```